lsif_indexer/indexer.py

```diff
@@ -0,0 +1,439 @@
+                model.Snippet(language="py", value=sig) for sig in signatures
+            ]
+        else:
+            hover_contents = [
+                model.Snippet(
+                    language="py", value=self.name.get_line_code().strip()
+                )
+            ]
+        if docstring := self.docstring():
+            hover_contents.append(docstring)
+        if hover_contents:
+            return model.HoverResult(
+                result=model.HoverResultContents(contents=hover_contents)
+            )
+        return None
+
+    def emit(self):
+        yield self.result_set
+        yield self.range
+        yield model.Next(out_v=self.range.id, in_v=self.result_set.id)
+
+        if hover_result := self.get_hover_result():
+            yield hover_result
+            yield model.Hover(out_v=self.result_set.id, in_v=hover_result.id)
+
+        if self.next_definition:
+            yield model.Next(
+                out_v=self.result_set.id,
+                in_v=self.next_definition.result_set.id,
+            )
+        else:
+            def_result = model.DefinitionResult()
+            yield def_result
+            yield model.Definition(
+                out_v=self.result_set.id, in_v=def_result.id
+            )
+            yield model.Item(
+                out_v=def_result.id,
+                in_vs=[self.range.id],
+                document=self.document_id,
+                property="definitions",
+            )
+
+            ref_result = model.ReferenceResult()
+            self._ref_result_id = ref_result.id
+            yield ref_result
+            yield model.Reference(out_v=self.result_set.id, in_v=ref_result.id)
+        self.add_reference(self.document_id, self.range.id)
+
+    def add_reference(self, document_id: int, range_id: int):
+        if self.next_definition:
+            self.next_definition.add_reference(document_id, range_id)
+        else:
+            self.references[document_id].append(range_id)
+
+    def emit_references(self):
+        for doc_id, ref_ids in self.references.items():
+            yield model.Item(
+                out_v=self._ref_result_id,
+                in_vs=ref_ids,
+                document=doc_id,
+                property="references",
+            )
+
+
+@dataclass
+class Reference(Range):
+    definition: Definition
+
+    def emit(self):
+        yield self.range
+        if self.document_id == self.definition.document_id:
+            result_set_id = self.definition.result_set.id
+        else:
+            result_set = model.ResultSet()
+            yield result_set
+            yield model.Next(
+                out_v=result_set.id, in_v=self.definition.result_set.id
+            )
+            result_set_id = result_set.id
+        yield model.Next(out_v=self.range.id, in_v=result_set_id)
+        self.definition.add_reference(self.document_id, self.range.id)
+
+
+DefinitionKey = tuple[Path, int, int]
```

Comment:
This could probably just be `tuple[int, int]`, I didn't end up having an index across all files, I have one index per file.